### ISSUE TITLE
Pods are in CrashLoopBackOff state in Hygon CPU environment.

### ISSUE THREAD

**Issue by Goend** [ceph/ceph-csi#4379]
# Describe the bug #
Pods always are in CrashLoopBackOff.some logs in dmesg.
However, this container image was built by me,
The community image is available in the environment.
I got it by 
```
make image-cephcsi
```
Later, I built it in GitHub Actions to avoid the difference with localhost.
```
---
name: multi-arch-build
# yamllint disable-line rule:truthy
on:
  pull_request:
    branches:
      - '*'
permissions:
  contents: read

jobs:
  codespell:
    name: multi-arch-build
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - name: multi-arch-build
        # yamllint disable-line rule:line-length
        if: ${{ ! contains(github.event.pull_request.labels.*.name, 'ci/skip/multi-arch-build') }}
        # podman cannot pull images with both tag and digest
        # https://github.com/containers/buildah/issues/1407
        # use docker to build images
        run: CONTAINER_CMD=docker ./scripts/build-multi-arch-image.sh
      - name: single-arch-build
        # yamllint disable-line rule:line-length
        if: ${{ contains(github.event.pull_request.labels.*.name, 'ci/skip/multi-arch-build') }}
        run: make containerized-build
      - name: Build binary
        run: |
          echo "creating images cache..."
          docker save hub.easystack.io/captain/cephcsi:v3.9.0-es| pigz > ceph-csi-x86.tar.gz
          docker save hub.easystack.io/arm64v8/cephcsi:v3.9.0-es| pigz > ceph-csi-arm.tar.gz
      - name: cache x86 image
        uses: actions/upload-artifact@v2
        with:
          name: ceph-csi-x86.tar.gz
          path: ceph-csi-x86.tar.gz
      - name: cache arm image
        uses: actions/upload-artifact@v2
        with:
          name: ceph-csi-arm.tar.gz
          path: ceph-csi-arm.tar.gz
```

# Environment details #
```
[root@node-6 ~]# lscpu
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              128
On-line CPU(s) list: 0-127
Thread(s) per core:  2
Core(s) per socket:  32
Socket(s):           2
NUMA node(s):        8
Vendor ID:           HygonGenuine
BIOS Vendor ID:      Chengdu Hygon
CPU family:          24
Model:               2
Model name:          Hygon C86 7380 32-core Processor
BIOS Model name:     Hygon C86 7380 32-core Processor
Stepping:            2
CPU MHz:             2471.242
BogoMIPS:            4400.19
Virtualization:      AMD-V
L1d cache:           32K
L1i cache:           64K
L2 cache:            512K
L3 cache:            8192K
NUMA node0 CPU(s):   0-7,64-71
NUMA node1 CPU(s):   8-15,72-79
NUMA node2 CPU(s):   16-23,80-87
NUMA node3 CPU(s):   24-31,88-95
NUMA node4 CPU(s):   32-39,96-103
NUMA node5 CPU(s):   40-47,104-111
NUMA node6 CPU(s):   48-55,112-119
NUMA node7 CPU(s):   56-63,120-127
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid amd_dcm aperfmperf pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb hw_pstate ssbd ibpb vmmcall fsgsbase bmi1 avx2 smep bmi2 rdseed adx smap clflushopt sha_ni xsaveopt xsavec xgetbv1 xsaves clzero irperf xsaveerptr arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif overflow_recov succor smca sme sev sev_es
```

- Image/version of Ceph CSI driver :
v3.9.0
- Helm chart version :
- Kernel version :
```
[root@node-6 ~]# uname -a
Linux node-6.domain.tld 4.18.0-372.19.1.es8_8.x86_64 #1 SMP Tue Jan 2 14:54:00 CST 2024 x86_64 x86_64 x86_64 GNU/Linux
```
- Mounter used for mounting PVC (for cephFS its `fuse` or `kernel`. for rbd its
  `krbd` or `rbd-nbd`) :
- Kubernetes cluster version :
- Ceph cluster version :

# Steps to reproduce #

1. Setup details: '...'
1. Deployment to trigger the issue '....'
1. See error

# Actual results #

Describe what happened

# Expected behavior #

# Logs #

```
Jan 15 20:34:36 node-5 kernel: cephcsi[20662]: segfault at 53666c0 ip 00007fa9381ecf20 sp 00007ffd196ea348 error 6 in libc-2.28.so[7fa9381c8000+1cd000]
Jan 15 20:34:52 node-5 kernel: cephcsi[20944]: segfault at 51af6c0 ip 00007f59a1728f20 sp 00007ffe99f6f968 error 6 in libc-2.28.so[7f59a1704000+1cd000]
Jan 15 20:35:18 node-5 kernel: cephcsi[21692]: segfault at 4e3f6c0 ip 00007efdb7fb3f20 sp 00007ffcd3b7e238 error 6 in libc-2.28.so[7efdb7f8f000+1cd000]
Jan 15 20:35:41 node-5 kernel: cephcsi[22239]: segfault at 39eb6c0 ip 00007fd8aef9af20 sp 00007ffe82e6e318 error 6 in libc-2.28.so[7fd8aef76000+1cd000]
Jan 15 20:35:42 node-5 kernel: cephcsi[22312]: segfault at 396c6c0 ip 00007f57c3e19f20 sp 00007ffc924a85e8 error 6 in libc-2.28.so[7f57c3df5000+1cd000]
Jan 15 20:35:55 node-5 kernel: cephcsi[22546]: segfault at 43226c0 ip 00007f973fcbef20 sp 00007ffcb73629a8 error 6 in libc-2.28.so[7f973fc9a000+1cd000]
Jan 15 20:36:21 node-5 kernel: cephcsi[22882]: segfault at 43ed6c0 ip 00007f9bc1f1ff20 sp 00007ffee4e8f888 error 6 in libc-2.28.so[7f9bc1efb000+1cd000]
Jan 15 20:37:02 node-5 kernel: cephcsi[23357]: segfault at 403a6c0 ip 00007f0ea1487f20 sp 00007ffe2518ad78 error 6 in libc-2.28.so[7f0ea1463000+1cd000]
Jan 15 20:38:23 node-5 kernel: cephcsi[24393]: segfault at 44f86c0 ip 00007f480207df20 sp 00007ffd68f94648 error 6 in libc-2.28.so[7f4802059000+1cd000]
```

**Comment by Goend** [comment]
If there are any problems with the way I build the image, please let me know. Thank you.

**Comment by Goend** [comment]
I found a way to skip this issue. I used the community image as the base image and only compiled and replaced the ceph csi binary file in it. It can run in the environment where the problem occurred. Since the difference may be caused by /lib64/libc.so.6, I would like to ask if it is a software source issue.

**Comment by Goend** [comment]
[logs_21.zip](https://github.com/ceph/ceph-csi/files/13959643/logs_21.zip)
Github Action logs

**Comment by nixpanic** [comment]
segfaults like that can be caused by compiling the executable on one os/version, and running it on an other. Ideally you compile and run the executable in containers with the same versions of all libraries/packages.

Did you use a different (base) container-image for building, compared to the container-image where the executable runs?

**Comment by Goend** [comment]
Base container-image: quay.io/ceph/ceph:v17
from github action logs
```
2024-01-17T04:11:56.1438144Z + IFS=

2024-01-17T04:11:56.1439370Z ++ awk -v ARCH=amd64 '{if (archfound) {print $NF; exit 0}}; {archfound=($0 ~ "arch.*"ARCH)}'

2024-01-17T04:11:56.1464184Z + digest='"sha256:c617b4ab6d51c605dd83c70f4e498c3815df7453dfd691fe2525614d1913949d"'

2024-01-17T04:11:56.1465235Z + IFS='   

2024-01-17T04:11:56.1465707Z '

2024-01-17T04:11:56.1467010Z + base_img='quay.io/ceph/ceph:v17@"sha256:c617b4ab6d51c605dd83c70f4e498c3815df7453dfd691fe2525614d1913949d"'

2024-01-17T04:11:56.1467850Z + build_step 'make image-cephcsi for amd64'

2024-01-17T04:11:56.1468369Z + echo 'make image-cephcsi for amd64'

2024-01-17T04:11:56.1469835Z ++ cat /tmp/tmp.AjtvN7KomE

2024-01-17T04:11:56.1480814Z + build_step_log 'starting: make image-cephcsi for amd64'

2024-01-17T04:11:56.1483300Z ++ date

2024-01-17T04:11:56.1494470Z + echo 'Wed Jan 17 04:11:56 UTC 2024 -- starting: make image-cephcsi for amd64'

2024-01-17T04:11:56.1495812Z Wed Jan 17 04:11:56 UTC 2024 -- starting: make image-cephcsi for amd64

2024-01-17T04:11:56.1496692Z + '[' -n 1730 ']'

2024-01-17T04:11:56.1497332Z + return

2024-01-17T04:11:56.1497780Z + GOARCH=amd64

2024-01-17T04:11:56.1499425Z + BASE_IMAGE='quay.io/ceph/ceph:v17@"sha256:c617b4ab6d51c605dd83c70f4e498c3815df7453dfd691fe2525614d1913949d"'

2024-01-17T04:11:56.1500361Z + make image-cephcsi
```
The operating system/kernel is the same version, the only difference is the hardware.
@nixpanic

**Comment by nixpanic** [comment]
If you build everything through the Ceph-CSI Makefile and build.env settings, then I expect that the executable and libraries in the container-image are all consistent. It does not explain why there would be segfaults.

It is not clear what you mean with this comment:

> I found a way to skip this issue. I used the community image as the base image and only compiled and replaced the ceph csi binary file in it. It can run in the environment where the problem occurred. Since the difference may be caused by /lib64/libc.so.6, I would like to ask if it is a software source issue.

Replacing the executable with one that was built in a different environment is more likely to cause problems...

Did you use a different Golang version to compile the executable? You may be able to inspect the executables and compare compilation flags or other (think of  things like `objdump --all-headers --info cephcsi`).

**Comment by Goend** [comment]
The community's makefile is used to compile Golang in a container, so there should be no version issues. 

You can check my GitHub Action logs. I have not changed anything except the final image tag. 

I want to ask if there are other conditions that could cause these libraries to be updated, results in 
```
ldd /usr/local/bin/cephcsi
```

**Comment by nixpanic** [comment]
ldd /usr/local/bin/cephcsi

is just a check that verifies all dynamically linked libraries are available. librados and other Ceph libraries (with dependencies) should be findable in the container-image, if `ldd` can not find them, the executable won't work.

**Comment by nixpanic** [comment]
Do you see segfaults with images from https://quay.io/repository/cephcsi/cephcsi?tab=tags too?

**Comment by Goend** [comment]
> Do you see segfaults with images from https://quay.io/repository/cephcsi/cephcsi?tab=tags too?

No,The community image is available in the environment.
Therefore, I am confused as to why the images produced by the local execution of the simplest commands 
```
make image-cephcsi 
```
and multi-arch-build pipeline don't work but the community images do.

Is there any other preparation here that I'm not aware of?

**Comment by nixpanic** [comment]
The image building and pushing to the registry is done through the workflows that you can find in `.github/workflows`. For the multi-arch images/manifest you can check the `deploy.sh` script in the root directory of the project too.

As a Hygon CPU should just be a standard x86-64 / amd64, you do not even need to bother with multi-arch images. Running `make image-cephcsi` on a Hygon system should just work :tm:

**Comment by Goend** [comment]
> The image building and pushing to the registry is done through the workflows that you can find in `.github/workflows`. For the multi-arch images/manifest you can check the `deploy.sh` script in the root directory of the project too.
> 
> As a Hygon CPU should just be a standard x86-64 / amd64, you do not even need to bother with multi-arch images. Running `make image-cephcsi` on a Hygon system should just work ™️


As you mentioned, the image will only have architectural differences. This does not explain why the image runs normally on Intel nodes but fails to start on Hygon nodes. This is because our images are built in a unified manner, and therefore cannot be tailored to specific cases, namely, building on Hygon nodes.
Intel node     vs   Hygon node
same operating system
same architecture

error code
```
00000000000a3d50 <__memcpy_sse2_unaligned_erms>:
   a3d50:       f3 0f 1e fa             endbr64
   a3d54:       48 89 f8                mov    %rdi,%rax
   a3d57:       48 83 fa 10             cmp    $0x10,%rdx
...
   a3f83:       4c 39 d6                cmp    %r10,%rsi
   a3f86:       0f 82 27 ff ff ff       jb     a3eb3 <__memcpy_sse2_unaligned_erms+0x163>
   a3f8c:       0f 18 8e 80 00 00 00    prefetcht0 0x80(%rsi)
   a3f93:       0f 18 8e c0 00 00 00    prefetcht0 0xc0(%rsi)
   a3f9a:       0f 10 06                movups (%rsi),%xmm0
   a3f9d:       0f 10 4e 10             movups 0x10(%rsi),%xmm1
   a3fa1:       0f 10 56 20             movups 0x20(%rsi),%xmm2
   a3fa5:       0f 10 5e 30             movups 0x30(%rsi),%xmm3
...
   a4048:       45 0f 11 03             movups %xmm8,(%r11)
   a404c:       c3                      retq
   a404d:       0f 1f 00                nopl   (%rax)
```


with problem:
libc-2.28-246.el8.6.x86_64.debug
no problem:
libc-2.28.so-2.28-236.el8.6.x86_64.debug

changes from 2.28-236 to 2.28-246 ：https://centos.pkgs.org/8-stream/centos-baseos-aarch64/glibc-2.28-246.el8.aarch64.rpm.html
2023-12-14 - DJ Delorie <dj@redhat.com> - 2.28-246
- Include CentOS Hyperscaler SIG patches backported by Intel (RHEL-15696)
2023-12-08 - Florian Weimer <fweimer@redhat.com> - 2.28-245
- Improve compatibility between underlinking and IFUNC resolvers (RHEL-16825)
2023-11-24 - Florian Weimer <fweimer@redhat.com> - 2.28-244
- Restore <sys/cdefs.h> compatibility with C90 compilers (RHEL-15867)
2023-11-21 - Florian Weimer <fweimer@redhat.com> - 2.28-243
- ldconfig should skip temporary files created by RPM (RHEL-13720)
2023-11-20 - Florian Weimer <fweimer@redhat.com> - 2.28-242
- Fix force-first handling in dlclose (RHEL-10481)
2023-11-10 - Florian Weimer <fweimer@redhat.com> - 2.28-241
- Avoid lazy binding failures during dlclose (RHEL-3639)
2023-10-24 - Arjun Shankar <arjun@redhat.com> - 2.28-240
- Add /usr/share/doc/glibc/gai.conf to glibc-doc (RHEL-12894)
2023-10-20 - Florian Weimer <fweimer@redhat.com> - 2.28-239
- nscd: Skip unusable entries in first pass in prune_cache (RHEL-1192)
2023-10-16 - DJ Delorie <dj@redhat.com> - 2.28-238
- Fix slow tls access after dlopen (RHEL-2122)
2023-10-16 - Arjun Shankar <arjun@redhat.com> - 2.28-237
- Enable running a single test from the testsuite (RHEL-3757)

How to modify to use the original libc version?

**Comment by nixpanic** [comment]
You should probably report this at the CentOS project as well. At least when you confirm that the previous version of libc works.

You can add a `dnf downgrade --yes glibc-2.28-236.el8.6` or similar in `deploy/cephcsi/image/Dockerfile` as part of the `updated_base` container.

**Comment by changlexc** [comment]
The bug was introduced from centos 8 stream which merged a imcompatible glibc version 2.28.251. Since then, all the images which base image is comming from centos 8 stream or the later release have this problem. To resolve the problem, you should build a new image which by using the host's /lib64 library. Or you can add -v to use the host /lib64 library dir.

Note: this is not a Hygon CPU instruction set problem, it's just glibc 2.28.251's problem. It's just on centos 8 stream, centos 8 or centos 9 don't has this problem.
